addons/binding/org.openhab.binding.smartthings/src/main/java/org/openhab/binding/smartthings/discovery/SmartthingsDiscoveryService.java

```diff
@@ -0,0 +1,244 @@
+        String data = (String) event.getProperty("data");
+        logger.debug("Event received on topic: {}", topic);
+
+        // Two classes are required.
+        // 1. SmarthingsDiscoveryData contains timing info and the discovery data which is sent as an array of Strings
+        // 2. SmartthingDeviceData contains the device data for one device.
+        // First the SmarthingsDiscoveryData is converted from json to java. Then each data string is converted into
+        // device data
+        SmartthingsDiscoveryData discoveryData = gson.fromJson(data, SmartthingsDiscoveryData.class);
+        long openHabStartTime = discoveryData.getOpenHabStartTime();
+        long hubTime = discoveryData.getHubTime();
+        long transmissionCompleteTime = System.currentTimeMillis();
+
+        for (String deviceStr : discoveryData.getData()) {
+            SmartthingsDeviceData deviceData = gson.fromJson(deviceStr, SmartthingsDeviceData.class);
+            createDevice(deviceData);
+        }
+
+        // Log processing time
+        logger.info(
+                "Discovery timing data, Request time until data received {}, Request time until data recieved and processed {}, Hub processing time: {} ",
+                transmissionCompleteTime - openHabStartTime, System.currentTimeMillis() - openHabStartTime, hubTime);
+    }
+
+    /**
+     * Create a device with the data from the Smartthings hub
+     *
+     * @param deviceData
+     */
+    private void createDevice(SmartthingsDeviceData deviceData) {
+        logger.info("Discovery: Creating device: ThingType {} with name {}", deviceData.getCapability(),
+                deviceData.getName());
+
+        // Build the UID as a string smartthings:{ThingType}:{BridgeName}:{DeviceName}
+        String deviceNameNoSpaces = deviceData.getName().replaceAll("\\s", "_");
+        String smartthingsDeviceName = findIllegalChars.matcher(deviceNameNoSpaces).replaceAll("");
+        ThingUID bridgeUid = bridgeHandler.getThing().getUID();
+        String bridgeId = bridgeUid.getId();
+        String uidStr = String.format("smartthings:%s:%s:%s", deviceData.getCapability(), bridgeId,
+                smartthingsDeviceName);
+
+        Map<String, Object> properties = new HashMap<>();
+        properties.put("smartthingsName", deviceData.getName());
+        properties.put("deviceId", deviceData.getId());
+
+        DiscoveryResult discoveryResult = DiscoveryResultBuilder.create(new ThingUID(uidStr)).withProperties(properties)
+                .withRepresentationProperty(deviceData.getId()).withBridge(bridgeUid).withLabel(deviceData.getName())
+                .build();
+
+        thingDiscovered(discoveryResult);
+    }
+
+    /**
+     * Scanning worker class.
+     */
+    public class SmartthingsScan implements Runnable {
+        /**
+         * Handler for delegation to callbacks.
+         */
+        private SmartthingsDiscoveryService service;
+
+        /**
+         * Constructor.
```

Comment:
If you don't have to say anything in a javadoc, leave it out. Think about your LOC's.